extendr-api/src/robj/into_robj.rs

```diff
@@ -39,16 +51,157 @@ impl From<()> for Robj {
         res.unwrap().into()
     }
 }
 
+/// Convert a Result to an Robj either an Ok value or the Err value wrapped in a
+/// error condition. This is used to allow functions to use the ? operator
+/// and return [Result<T>] without panicking on an Err. T must impl IntoRobj.
+///
+/// Returns ok-value as is. Return err wrapped in a R error condition. The err is placed in
+/// $value. The condition messeage is simply 'extendr_arr'
+/// ```
+/// use extendr_api::prelude::*;
+/// fn my_func() -> Result<f64> {
+///     Ok(1.0)
+/// }
+///
+/// test! {
+///     assert_eq!(r!(my_func()), r!(1.0));
+/// }
+///
+/// //ok and err type is any IntoRobj
+/// fn my_err_f() -> std::result::Result<f64, f64> {
+///     Err(42.0) // return err float
+/// }
+///
+/// test! {
+///     assert_eq!(
+///         r!(my_err_f()),
+///         R!(
+/// "structure(list(message = 'extendr_err',
+/// value = 42.0), class = c('extendr_error', 'error', 'condition'))"
+///         ).unwrap()
+///     );
+/// }
+///
+/// ```
+#[cfg(all(feature = "result_condition", not(feature = "result_list")))] //write this feature, right any feature with precedence
+impl<T, E> From<std::result::Result<T, E>> for Robj
+where
+    T: Into<Robj>,
+    E: Into<Robj>,
+{
+    fn from(res: std::result::Result<T, E>) -> Self {
+        match res {
+            Ok(x) => x.into(),
+            Err(x) => {
+                let robj: Robj = x.into();
+                List::from_names_and_values(
+                    &["message", "value"],
+                    &["extendr_err".into_robj(), robj],
+                )
+            }
+            //can only imagine this would ever fail due memory allcation error, but then panicking is the right choice
+            .expect("internal error: failed to create an R list")
+            .set_class(["extendr_error", "error", "condition"])
+            .expect("internal error: failed to set class"),
+        }
+    }
+}
+
+/// Convert a Result to an R `List` with an `ok` and `err` element.
+/// This is used to allow functions to use the ? operator
+/// and return [std::result::Result<T,E> or extendr_api::result::Result<T>]
+/// without panicking on an Err.
+///
+///
+/// ```
+/// use extendr_api::prelude::*;
+/// fn my_err_f() -> std::result::Result<f64, String> {
+///     Err("We have water in the engine room!".to_string())
+/// }
+/// fn my_ok_f() -> std::result::Result<f64, String> {
+///     Ok(123.123)
+/// }
+///
+/// test! {
+///     assert_eq!(
+///         r!(my_err_f()),
+///         R!("x=list(ok=NULL, err='We have water in the engine room!')
+///             class(x)='extendr_result'
+///             x"
+///         ).unwrap()
+///     );
+///     assert_eq!(
+///         r!(my_ok_f()),
+///         R!("x = list(ok=123.123, err=NULL)
+///             class(x)='extendr_result'
+///             x"
+///         ).unwrap()
+///     );
+/// }
+///
+/// ```
+#[cfg(feature = "result_list")] //write this feature and not any result_feature with presecence
```

Comment:
Same here, please clarify the comment

Reply:
comment dropped

The comment was an instruction for how to add future encoding features. I don't think it was necessary.